### PR TITLE
riscv_common: Refactor common fe310 code to riscv_common

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This PR refactors the lone fe310 RISC-V cpu into a separate riscv_common director keeping the fe310-specifics in their own directory. This makes it easier to add new RISC-V based processors in the future.

My original idea was to refactor into a rv32i_common directory, but I decided that riscv_common should be enough. Changes between rv64i and rv32i are minimal and mostly in the architecture size and doesn't warrant separate directories for the two (for now). It is however possible that some doxygen or variable names are still `rv32i_`, these should all be renamed to `riscv_`

### Testing procedure

Changes to the binaries should be minimal or nonexistent.

Compilation from Murdock, together with a few applications tested on the hifive1b should be enough for code tests.
Please also verify that the documentation and the generated doxygen makes sense after the split.

### Issues/PRs references

None

### CODE REVIEW

File: cpu/riscv_common/include/periph_cpu_common.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * @name    Power management configuration
+ * @{
+ */
+#define PROVIDES_PM_SET_LOWEST
+/** @} */
+
+/**
+ * @brief   Length of the CPU_ID in octets
+ */
+#define CPUID_LEN           (12U)
```
Comment (benpicco) [inline cpu/riscv_common/include/periph_cpu_common.h:40]: Is this true for all RISC-V CPUs?
Or should that rather be moved to fe310

Comment (bergzand) [reply cpu/riscv_common/include/periph_cpu_common.h:40]: Should probably be moved to the fe310, there is no common CPU_ID register on the RISC-V architecture. For the fe310, the vendor ID, the architecture ID and the implementation ID are concatenated. Not perfect and probably identical between different boards, but I haven't found a better ID in the mean time.

I'll move this to the fe310 header

File: cpu/riscv_common/include/cpu.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * @brief   Enable the rv32i FPU when present
+ */
+void riscv_fpu_init(void);
+
+/**
+ * @brief   Initialization of the interrupt controller
+ */
+void riscv_irq_init(void);
```
Comment (benpicco) [inline cpu/riscv_common/include/cpu.h:52]: should those be public?

**Review by benpicco — COMMENTED** [review]
looks good to me, but I don't have any hardware to test this.

**Comment by bergzand** [comment]
There are still a few places where the riscv_common code uses code from the fe310. Slapping on a WIP label to prevent accidental merges while this is still the case.

File: cpu/riscv_common/ldscripts/riscv.ld [hunk 3]
```diff
@@ -7,24 +7,24 @@
  */
 
 /**
- * @addtogroup      cpu_fe310
+ * @addtogroup         cpu_riscv_common
  * @{
  *
  * @file
- * @brief           Memory definitions for the SiFive FE310
+ * @brief           Memory definitions for for the RISC-V CPU
```
Comment (aabadie) [inline cpu/riscv_common/ldscripts/riscv.ld:14]:
```suggestion
 * @brief           Memory definitions for the RISC-V CPU
```

**Comment by bergzand** [comment]
TODO:

 - Migrate _REG32(p, i) defines to a common place and allow periph implementations to define PERIPH_REG defines
 - Migrate PLIC config defines to a periph_cpu header: `PLIC_CTRL_ADDR` `PLIC_NUM_INTERRUPTS` and `PLIC_NUM_PRIORITIES`
 - Allow setting a `TARGET_ARCH_MARCH` string (rv32imac` and a `TARGET_ARCH_MABI` (ilp32)

File: cpu/fe310/include/cpu_conf.h [hunk 4]
```diff
@@ -19,35 +19,15 @@
...
- */
-#define IRQ_API_INLINED     (1)
+#include "cpu_conf_common.h"
+#include "vendor/platform.h"
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+#define CLINT_BASE_ADDR     (CLINT_CTRL_ADDR)
```
Comment (aabadie) [inline cpu/fe310/include/cpu_conf.h:29]: I think this define should be documented

**Comment by aabadie** [comment]
Without the following patch, the build is broken for application requiring irq/timer:

```diff
index c66b62719e..2582e92261 100644
--- a/cpu/riscv_common/Makefile.dep
+++ b/cpu/riscv_common/Makefile.dep
@@ -4,6 +4,8 @@ USEMODULE += riscv_common
 # include common periph code
 USEMODULE += riscv_periph_common
 
+FEATURES_REQUIRED += periph_coretimer periph_plic
+
 USEMODULE += periph_pm
 
 # Make calls to malloc and friends thread-safe
diff --git a/cpu/riscv_common/Makefile.features b/cpu/riscv_common/Makefile.features
index 6880670a79..f0eca40db5 100644
--- a/cpu/riscv_common/Makefile.features
+++ b/cpu/riscv_common/Makefile.features
@@ -6,6 +6,7 @@ FEATURES_PROVIDED += periph_cpuid
 FEATURES_PROVIDED += periph_plic
 FEATURES_PROVIDED += periph_pm
 FEATURES_PROVIDED += ssp
+FEATURES_PROVIDED += periph_coretimer periph_plic
 
 # RISC-V toolchain on CI does not work properly with picolibc yet
 ifeq (,$(RIOT_CI_BUILD))
```

I'm not sure that declaring the coretimer/plic as peripherals is the way to go. You could simply add them to the riscv_common module. But maybe I'm missing something here ?

**Comment by fjmolinas** [comment]
Is this still wip @bergzand ?

**Comment by bergzand** [comment]
> Is this still wip @bergzand ?

Unfortunately

**Comment by bergzand** [comment]
> I'm not sure that declaring the coretimer/plic as peripherals is the way to go. You could simply add them to the riscv_common module. But maybe I'm missing something here ?

Me neither. The issue is that the PLIC is one of the possible implementations for interrupt controllers on the RISC-V. There is also the [CLIC](https://github.com/riscv/riscv-fast-interrupt/blob/master/clic.adoc). A CPU could use the CLINT timer for periph_timer, but it could also use a different (more advanced) timer peripheral

File: cpu/fe310/include/cpu_conf.h (hunk 4 above)
Comment (bergzand) [reply cpu/fe310/include/cpu_conf.h:32]: I've added a bit of doxygen. More extensive documentation is available in the `periph/coretimer.c` file

File: cpu/riscv_common/include/cpu.h (hunk 2 above)
Comment (bergzand) [reply cpu/riscv_common/include/cpu.h:52]: Doesn't have to be, but I'm going to leave that cleanup to a possible follow up :)

**Comment by bergzand** [comment]
@fjmolinas No longer WIP!

File: cpu/fe310/include/periph_cpu.h [hunk 5]
```diff
@@ -67,9 +61,9 @@ typedef uint8_t gpio_t;
  */
 typedef struct {
     uint32_t addr;          /**< UART control register address */
-    gpio_t rx;              /**< RX pin */
-    gpio_t tx;              /**< TX pin */
-    plic_source isr_num;    /**< ISR source number */
+    gpio_t   rx;              /**< RX pin */
+    gpio_t   tx;              /**< TX pin */
+    irqn     isr_num;    /**< ISR source number */
```
Comment (aabadie) [inline cpu/fe310/include/periph_cpu.h:66]: Please keep the doxygen comment aligned:
```suggestion
    gpio_t   rx;            /**< RX pin */
    gpio_t   tx;            /**< TX pin */
    irqn     isr_num;       /**< ISR source number */
```

File: cpu/riscv_common/Makefile.dep [hunk 6]
```diff
@@ -0,0 +1,10 @@
+# Tell the build system that the CPU depends on the risc-v common files:
+USEMODULE += riscv_common
+
+# include common periph code
+USEMODULE += riscv_periph_common
+
+USEMODULE += periph_pm
```
Comment (aabadie) [inline cpu/riscv_common/Makefile.dep:7]: I think this line is not needed because the periph_pm feature is always added if provided by the platform:

https://github.com/RIOT-OS/RIOT/blob/38188017a810105b67a3771735287c34d3fdba42/Makefile.dep#L89-L09

 And the FEATURES_REQUIRED is turned into a USEMODULE:

https://github.com/RIOT-OS/RIOT/blob/38188017a810105b67a3771735287c34d3fdba42/makefiles/features_modules.inc.mk#L7-L10

So you can drop this line:

```suggestion
```

File: cpu/riscv_common/include/periph_cpu_common.h [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+ * @name    Power management configuration
+ * @{
+ */
+#define PROVIDES_PM_SET_LOWEST
+/** @} */
+
+/**
+ * @brief Convenience typedef for interrupt lines
+ */
+typedef unsigned irqn;
```
Comment (aabadie) [inline cpu/riscv_common/include/periph_cpu_common.h:40]: In RIOT the type are usually suffixed with `_t`. Better use `irqn_t` I think:

```suggestion
typedef unsigned irqn_t;
```

Comment (bergzand) [reply cpu/riscv_common/include/periph_cpu_common.h:40]: I agree. Unrelated to this PR though.

File: cpu/riscv_common/Makefile.dep (hunk 6 above)
Comment (bergzand) [reply cpu/riscv_common/Makefile.dep:7]: Thanks, removed!

File: cpu/fe310/include/periph_cpu.h (hunk 5 above)
Comment (bergzand) [reply cpu/fe310/include/periph_cpu.h]: Aligned them

File: cpu/riscv_common/include/periph_cpu_common.h (hunk 7 above)
Comment (aabadie) [reply cpu/riscv_common/include/periph_cpu_common.h:40]: > Unrelated to this PR though.

Are you sure ? This change is introduced by this PR (no existing code reported by `git grep "irqn" cpu/fe310` in master)

Comment (bergzand) [reply cpu/riscv_common/include/periph_cpu_common.h:40]: > Are you sure ?

Not anymore, maybe I was too quick. I'll check it and fix if I indeed introduce it here.

Comment (bergzand) [reply cpu/riscv_common/include/periph_cpu_common.h:40]: Technically it is a rename from `plic_source`. I'll append the `_t` to the typedef.

Comment (bergzand) [reply cpu/riscv_common/include/periph_cpu_common.h]: Resolved!

File: cpu/riscv_common/periph/plic.c [hunk 8]
```diff
@@ -65,7 +74,7 @@ void plic_disable_interrupt(unsigned irq)
 {
     volatile uint32_t *irq_reg = _get_irq_reg(irq);
 
-    __atomic_fetch_and(irq_reg, ~(1 << (irq & 0x1f)), __ATOMIC_RELAXED);
+     __atomic_fetch_and(irq_reg, ~(1 << (irq & 0x1f)), __ATOMIC_RELAXED);
```
Comment (aabadie) [inline cpu/riscv_common/periph/plic.c:77]: Seems like an unrelated change
```suggestion
    __atomic_fetch_and(irq_reg, ~(1 << (irq & 0x1f)), __ATOMIC_RELAXED);
```

**Comment by aabadie** [comment]
Please squash, I'd like to know what Murdock has to say with this PR :)

**Comment by aabadie** [comment]
This needs a rebase now.

**Comment by bergzand** [comment]
Rebased!

**Comment by fjmolinas** [comment]
This one needs a rebase again @bergzand

**Comment by bergzand** [comment]
Rebased!

**Review by aabadie — CHANGES_REQUESTED** [review]
Still needs some work. With the following diff, things should be better:

```diff
diff --git a/cpu/fe310/Makefile b/cpu/fe310/Makefile
index 0dbf0f1807..fd024edf0e 100644
--- a/cpu/fe310/Makefile
+++ b/cpu/fe310/Makefile
@@ -2,6 +2,6 @@
 MODULE = cpu
 
 # add a list of subdirectories, that should also be built
-DIRS += $(RIOTCPU)/riscv_common periph vendor
+DIRS = $(RIOTCPU)/riscv_common periph vendor
 
-include $(RIOTCPU)/riscv_common/Makefile
+include $(RIOTBASE)/Makefile.base
diff --git a/cpu/fe310/Makefile.features b/cpu/fe310/Makefile.features
index 5715137124..66120f1516 100644
--- a/cpu/fe310/Makefile.features
+++ b/cpu/fe310/Makefile.features
:...skipping...
diff --git a/cpu/fe310/Makefile b/cpu/fe310/Makefile
index 0dbf0f1807..fd024edf0e 100644
--- a/cpu/fe310/Makefile
+++ b/cpu/fe310/Makefile
@@ -2,6 +2,6 @@
 MODULE = cpu
 
 # add a list of subdirectories, that should also be built
-DIRS += $(RIOTCPU)/riscv_common periph vendor
+DIRS = $(RIOTCPU)/riscv_common periph vendor
 
-include $(RIOTCPU)/riscv_common/Makefile
+include $(RIOTBASE)/Makefile.base
diff --git a/cpu/fe310/Makefile.features b/cpu/fe310/Makefile.features
index 5715137124..66120f1516 100644
--- a/cpu/fe310/Makefile.features
+++ b/cpu/fe310/Makefile.features
@@ -4,5 +4,6 @@ FEATURES_PROVIDED += cpp
 FEATURES_PROVIDED += libstdcpp
 FEATURES_PROVIDED += newlib
 FEATURES_PROVIDED += periph_cpuid
+FEATURES_PROVIDED += periph_gpio periph_gpio_irq
 
 include $(RIOTCPU)/riscv_common/Makefile.features
diff --git a/cpu/fe310/cpu.c b/cpu/fe310/cpu.c
index a244c1f49b..a10ea94dd6 100644
--- a/cpu/fe310/cpu.c
+++ b/cpu/fe310/cpu.c
@@ -18,7 +18,6 @@
  */
 
 #include "cpu.h"
-#include "nanostubs.h"
 #include "periph/init.h"
 #include "periph_conf.h"
 
diff --git a/cpu/riscv_common/Makefile b/cpu/riscv_common/Makefile
index 880129078d..e09377cd1e 100644
--- a/cpu/riscv_common/Makefile
+++ b/cpu/riscv_common/Makefile
@@ -1,3 +1,3 @@
-DIRS += periph
+DIRS = periph
 
 include $(RIOTBASE)/Makefile.base
diff --git a/cpu/riscv_common/Makefile.dep b/cpu/riscv_common/Makefile.dep
index 4c9bbe5ef7..70aa523b8c 100644
--- a/cpu/riscv_common/Makefile.dep
+++ b/cpu/riscv_common/Makefile.dep
@@ -2,7 +2,7 @@
 USEMODULE += riscv_common
 
 # include common periph code
-USEMODULE += riscv_periph_common
+USEMODULE += riscv_common_periph
 
 # Make calls to malloc and friends thread-safe
 USEMODULE += malloc_thread_safe
diff --git a/cpu/riscv_common/periph/Makefile b/cpu/riscv_common/periph/Makefile
index b83a4e8257..db019f82ea 100644
--- a/cpu/riscv_common/periph/Makefile
+++ b/cpu/riscv_common/periph/Makefile
@@ -1,2 +1,2 @@
-MODULE = riscv_periph_common
+MODULE = riscv_common_periph
 include $(RIOTMAKE)/periph.mk
```

File: cpu/fe310/Makefile [hunk 9]
```diff
@@ -2,6 +2,6 @@
 MODULE = cpu
 
 # add a list of subdirectories, that should also be built
-DIRS = periph vendor
+DIRS += $(RIOTCPU)/riscv_common periph vendor
 
-include $(RIOTBASE)/Makefile.base
+include $(RIOTCPU)/riscv_common/Makefile
```
[the remark this answers is not in the document]
Comment (bergzand) [reply cpu/fe310/Makefile:7]: I think you're right

**Comment by bergzand** [comment]
@aabadie Murdock is all green, do you mind if I squash here again?

**Comment by aabadie** [comment]
> do you mind if I squash here again?

No problem, go ahead!

**Comment by bergzand** [comment]
Squashed!

**Comment by aabadie** [comment]
I ran compile_and_test_for_board on this PR and there are a lot of test failing. Some are unrelated to this PR and are a regression compared to 2020.10.
I'll report here later.

**Comment by aabadie** [comment]
Here is the output of compile_and_test_for_board run on hifive1b with this PR:

```
ERROR:hifive1b:Tests failed: 10
Failures during test:
- [tests/event_wait_timeout](tests/event_wait_timeout/test.failed)
- [tests/gnrc_ipv6_fwd_w_sub](tests/gnrc_ipv6_fwd_w_sub/test.failed)
- [tests/periph_wdt](tests/periph_wdt/test.failed)
- [tests/pkg_utensor](tests/pkg_utensor/test.failed)
- [tests/ps_schedstatistics](tests/ps_schedstatistics/test.failed)
- [tests/sys_architecture](tests/sys_architecture/test.failed)
- [tests/thread_flags](tests/thread_flags/test.failed)
- [tests/thread_flags_xtimer](tests/thread_flags_xtimer/test.failed)
- [tests/xtimer_periodic_wakeup](tests/xtimer_periodic_wakeup/test.failed)
- [tests/xtimer_usleep](tests/xtimer_usleep/test.failed)
```

As said above, some of the failing tests are unrelated to this PR and some are a regression introduced by #15736 (unfortunately in the latest release), so I compared them against master (just re-ran the failing tests with the master branch):

```
ERROR:hifive1b:Tests failed: 10
Failures during test:
- [tests/event_wait_timeout](tests/event_wait_timeout/test.failed)
- [tests/gnrc_ipv6_fwd_w_sub](tests/gnrc_ipv6_fwd_w_sub/test.failed)
- [tests/periph_wdt](tests/periph_wdt/test.failed)
- [tests/pkg_utensor](tests/pkg_utensor/test.failed)
- [tests/ps_schedstatistics](tests/ps_schedstatistics/test.failed)
- [tests/sys_architecture](tests/sys_architecture/test.failed)
- [tests/thread_flags](tests/thread_flags/test.failed)
- [tests/thread_flags_xtimer](tests/thread_flags_xtimer/test.failed)
- [tests/xtimer_periodic_wakeup](tests/xtimer_periodic_wakeup/test.failed)
- [tests/xtimer_usleep](tests/xtimer_usleep/test.failed)
```

As a reference, here are results compared to 2021.01-devel:

```
ERROR:hifive1b:Tests failed: 7
Failures during test:
- [tests/gnrc_ipv6_fwd_w_sub](tests/gnrc_ipv6_fwd_w_sub/test.failed)
- [tests/periph_wdt](tests/periph_wdt/test.failed)
- [tests/pkg_utensor](tests/pkg_utensor/test.failed)
- [tests/ps_schedstatistics](tests/ps_schedstatistics/test.failed)
- [tests/sys_architecture](tests/sys_architecture/test.failed)
- [tests/xtimer_periodic_wakeup](tests/xtimer_periodic_wakeup/test.failed)
- [tests/xtimer_usleep](tests/xtimer_usleep/test.failed)
```

So #15736 introduced a regression for `tests/event_wait_timeout`, `tests/thread_flags` and `tests/thread_flags_xtimer`.

**Comment by bergzand** [comment]
So all failures are unrelated to this PR?

**Review by aabadie — APPROVED** [review]
Compared to master, there's no regression introduced by this PR. I added a https://github.com/RIOT-OS/RIOT/issues/13086#issuecomment-774999082 in #13086 regarding the regression I found in master, so we don't loose track of them.

Anyway, let's not block this PR because of them.

I'm fine with the changes here, they all look good.

ACK

**Comment by bergzand** [comment]
Thanks for the review! I'll try to pick up the regression asap
